src/storage/garbage_collector.cpp

```diff
@@ -7,109 +7,37 @@
 #include "transaction/deferred_action_manager.h"
 #include "transaction/transaction_context.h"
 #include "transaction/transaction_defs.h"
-#include "transaction/transaction_manager.h"
 #include "transaction/transaction_util.h"
 
 namespace terrier::storage {
 
-std::pair<uint32_t, uint32_t> GarbageCollector::PerformGarbageCollection() {
+void GarbageCollector::CleanupTransaction(transaction::TransactionContext *txn) {
   if (observer_ != nullptr) observer_->ObserveGCInvocation();
-  timestamp_manager_->CheckOutTimestamp();
-  const transaction::timestamp_t oldest_txn = timestamp_manager_->OldestTransactionStartTime();
-  uint32_t txns_deallocated = ProcessDeallocateQueue(oldest_txn);
-  STORAGE_LOG_TRACE("GarbageCollector::PerformGarbageCollection(): txns_deallocated: {}", txns_deallocated);
-  uint32_t txns_unlinked = ProcessUnlinkQueue(oldest_txn);
-  STORAGE_LOG_TRACE("GarbageCollector::PerformGarbageCollection(): txns_unlinked: {}", txns_unlinked);
-  if (txns_unlinked > 0) {
-    // Only update this field if we actually unlinked anything, otherwise we're being too conservative about when it's
-    // safe to deallocate the transactions in our queue.
-    last_unlinked_ = timestamp_manager_->CheckOutTimestamp();
+  if (txn->IsReadOnly()) {
+    // This is a read-only transaction so this is safe to immediately delete
+    delete txn;
+  } else {
+    timestamp_manager_->CheckOutTimestamp();
+    const transaction::timestamp_t oldest_txn = timestamp_manager_->OldestTransactionStartTime();
```

Comment:
I don't think we can call this on a per-txn basis (see: @GustavoAngulo's warning on the function). It takes the latch every time. It was designed to be done once per GC wakeup. Deferred action manager already calls it in Process(), so it updates the cached value. The actions should be reading the cached value.

I'm wondering if we need to rename that method to `AdvanceEpoch` or something, and then have `CachedEpoch`.